File: windows/src/discord_ipc.rs
//! A hand-written Discord RPC IPC client. No third-party RPC dependencies —
//! only serde for JSON, mirroring the macOS app's zero-dependency stance.
//!
//! Port of the transport in `AgentCord/DiscordIPC.swift`. The protocol is
//! byte-for-byte identical; only the transport differs:
//!
//!   * macOS  — Unix domain socket at `$TMPDIR/discord-ipc-{0..9}`
//!   * Windows — named pipe at `\\.\pipe\discord-ipc-{0..9}`
//!
//! On Windows a named pipe is just a file handle, so we open it with the
//! standard library's `OpenOptions` and use `Read`/`Write` directly.
//!
//! This module is the low-level toolkit: pipe discovery and frame read/write.
//! The connection lifecycle (handshake, reconnect-with-backoff, ping/pong) is
//! orchestrated by `presence_controller`, the way `PresenceController.swift`
//! drives `DiscordIPC` on macOS.
//!
//! Frame format on the wire:
//!   [ opcode: u32 LE ][ payloadLength: u32 LE ][ JSON bytes ]

use std::fs::{File, OpenOptions};
use std::io::{self, Read, Write};
use std::sync::atomic::{AtomicU64, Ordering};

use crate::models::{HandshakePayload, RichPresence, SetActivityCommand};

/// Discord IPC opcodes.
pub mod opcode {
    pub const HANDSHAKE: u32 = 0;
    pub const FRAME: u32 = 1;
    pub const CLOSE: u32 = 2;
    pub const PING: u32 = 3;
    pub const PONG: u32 = 4;
}

/// Open the first available `\\.\pipe\discord-ipc-{0..9}` for read+write.
pub fn open_pipe() -> Option<File> {
    (0..=9).find_map(|i| {
        OpenOptions::new()
            .read(true)
            .write(true)
            .open(format!(r"\\.\pipe\discord-ipc-{i}"))
            .ok()
    })
}

/// Encode the opcode-0 handshake payload for a client id.
pub fn handshake_payload(client_id: &str) -> Vec<u8> {
    serde_json::to_vec(&HandshakePayload { v: 1, client_id })
        .expect("HandshakePayload serialization is infallible")
}

/// Write one framed message: `[opcode LE][len LE][payload]`.
pub fn write_frame(w: &mut impl Write, opcode: u32, payload: &[u8]) -> io::Result<()> {
    let mut frame = Vec::with_capacity(8 + payload.len());
    frame.extend_from_slice(&opcode.to_le_bytes());
    frame.extend_from_slice(&(payload.len() as u32).to_le_bytes());
    frame.extend_from_slice(payload);
    w.write_all(&frame)?;
    w.flush()
}

/// Read one frame. `Ok(None)` means the peer closed the pipe (reconnect).
pub fn read_frame(r: &mut impl Read) -> io::Result<Option<(u32, Vec<u8>)>> {
    let mut header = [0u8; 8];
    if let Err(e) = r.read_exact(&mut header) {
        return match e.kind() {
            io::ErrorKind::UnexpectedEof | io::ErrorKind::BrokenPipe => Ok(None),
            _ => Err(e),
        };
    }
    let opcode = u32::from_le_bytes([header[0], header[1], header[2], header[3]]);
    let length = u32::from_le_bytes([header[4], header[5], header[6], header[7]]) as usize;
    if length == 0 {
        return Ok(Some((opcode, Vec::new())));
    }
    let mut payload = vec![0u8; length];
    if let Err(e) = r.read_exact(&mut payload) {
        return match e.kind() {
            io::ErrorKind::UnexpectedEof | io::ErrorKind::BrokenPipe => Ok(None),
            _ => Err(e),
        };
    }
    Ok(Some((opcode, payload)))
}

/// Result of handling one inbound IPC frame.
#[derive(Debug, PartialEq, Eq)]
pub enum FrameAction {
    /// Keep reading; no terminal state change.
    Continue,
    /// Discord sent READY — handshake complete.
    Ready,
    /// Discord sent ERROR with a message.
    Error(String),
    /// Peer closed the connection.
    Closed,
}

/// Handle one inbound frame: answer PINGs, detect READY/ERROR/CLOSE.
pub fn handle_inbound_frame<W: Write>(op: u32, payload: &[u8], pipe: &mut W) -> FrameAction {
    match op {
        opcode::FRAME => {
            let Ok(v) = serde_json::from_slice::<serde_json::Value>(payload) else {
                return FrameAction::Continue;
            };
            match v.get("evt").and_then(|e| e.as_str()) {
                Some("READY") => FrameAction::Ready,
                Some("ERROR") => {
                    let msg = v
                        .get("data")
                        .and_then(|d| d.get("message"))
                        .and_then(|m| m.as_str())
                        .unwrap_or("Discord reported an error")
                        .to_string();
                    FrameAction::Error(msg)
                }
                _ => FrameAction::Continue,
            }
        }
        opcode::PING => {
            let _ = write_frame(pipe, opcode::PONG, payload);
            FrameAction::Continue
        }
        opcode::CLOSE => FrameAction::Closed,
        _ => FrameAction::Continue,
    }
}

/// Read frames until READY. Answers PINGs along the way.
pub fn wait_for_ready(pipe: &mut (impl Read + Write)) -> Result<(), String> {
    loop {
        match read_frame(pipe) {
            Ok(Some((op, payload))) => match handle_inbound_frame(op, &payload, pipe) {
                FrameAction::Ready => return Ok(()),
                FrameAction::Error(msg) => return Err(msg),
                FrameAction::Closed => return Err("connection closed".to_string()),
                FrameAction::Continue => {}
            },
            Ok(None) => return Err("connection closed".to_string()),
            Err(e) => return Err(e.to_string()),
        }
    }
}

/// Open the pipe, send the handshake, and block until READY.
pub fn connect_handshake(client_id: &str) -> io::Result<File> {
    let mut pipe = open_pipe()
        .ok_or_else(|| io::Error::new(io::ErrorKind::NotConnected, "no discord-ipc pipe found"))?;
    write_frame(&mut pipe, opcode::HANDSHAKE, &handshake_payload(client_id))?;
    wait_for_ready(&mut pipe).map_err(|msg| io::Error::new(io::ErrorKind::NotConnected, msg))?;
    Ok(pipe)
}

/// Build and write a `SET_ACTIVITY` frame (or, with `None`, a clear). The nonce
/// counter is shared so every command gets a unique nonce.
pub fn write_activity(
    w: &mut impl Write,
    pid: u32,
    nonce: &AtomicU64,
    activity: Option<RichPresence>,
) -> io::Result<()> {
    let nonce_str = format!("{}-{}", pid, nonce.fetch_add(1, Ordering::Relaxed));
    let cmd = SetActivityCommand::new(nonce_str, pid, activity);
    let payload = serde_json::to_vec(&cmd).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
    write_frame(w, opcode::FRAME, &payload)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Cursor;

    #[test]
    fn ready_frame_returns_ready() {
        let payload = br#"{"evt":"READY"}"#;
        let mut buf: Vec<u8> = Vec::new();
        let action = handle_inbound_frame(opcode::FRAME, payload, &mut buf);
        assert_eq!(action, FrameAction::Ready);
    }

    #[test]
    fn ping_writes_pong() {
        let payload = b"1234";
        let mut written = Vec::new();
        let action = handle_inbound_frame(opcode::PING, payload, &mut written);
        assert_eq!(action, FrameAction::Continue);
        assert!(!written.is_empty());
    }

    #[test]
    fn error_frame_surfaces_message() {
        let payload = br#"{"evt":"ERROR","data":{"message":"bad id"}}"#;
        let mut buf = Cursor::new(Vec::new());
        let action = handle_inbound_frame(opcode::FRAME, payload, &mut buf);
        assert_eq!(action, FrameAction::Error("bad id".to_string()));
    }

    #[test]
    fn close_frame_is_closed() {
        let action = handle_inbound_frame(opcode::CLOSE, &[], &mut Vec::<u8>::new());
        assert_eq!(action, FrameAction::Closed);
    }
}

File: windows/DiscordIpc.cs
// A hand-written Discord RPC IPC client. No third-party dependencies.
// Implements only the subset needed for Rich Presence: pipe discovery,
// handshake, SET_ACTIVITY, and clearing. Port of AgentCord/DiscordIPC.swift.
//
// Transport: macOS uses a Unix domain socket at $TMPDIR/discord-ipc-{0..9};
// Windows uses a named pipe at \\.\pipe\discord-ipc-{0..9}. Opened with
// NamedPipeClientStream and PipeOptions.Asynchronous, so it uses overlapped
// I/O: a concurrent read loop and writer are safe — matching the macOS design.
// (A synchronous pipe would serialize reads and writes on one file object and
// deadlock a blocking reader against a writer, forcing a write-only loop.)
//
// Frame format on the wire:
//   [ opcode: UInt32 LE ][ payloadLength: UInt32 LE ][ JSON bytes ]

using System.IO;
using System.IO.Pipes;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace AgentCord;

public sealed class DiscordIpc : IDisposable
{
    public enum ConnState { Disconnected, Connecting, Connected }

    private enum Opcode : uint { Handshake = 0, Frame = 1, Close = 2, Ping = 3, Pong = 4 }

    // Callbacks may fire on background threads; UI code must marshal itself.
    public event Action<ConnState>? StateChanged;
    public event Action<string>? Error;
    public event Action? Ready;

    public ConnState State { get; private set; } = ConnState.Disconnected;

    private static readonly JsonSerializerOptions WireOptions = new()
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    private readonly object _lock = new();
    private readonly SemaphoreSlim _writeLock = new(1, 1);
    private readonly int _pid = Environment.ProcessId;

    private string _clientId = "";
    private bool _shouldRun;
    private bool _ready;
    private NamedPipeClientStream? _pipe;
    private CancellationTokenSource? _cts;
    private Task? _loop;

    /// <summary>The last non-null activity we were asked to display, re-sent on
    /// reconnect. Stored as its serialized JSON node for cheap dedup.</summary>
    private JsonNode? _currentActivity;
    private string? _sentSignature;

    // --- Public API

    /// <summary>Begin connecting (and keep reconnecting) using the given
    /// application id. Idempotent while already running with the same id.</summary>
    public void Connect(string clientId)
    {
        lock (_lock)
        {
            if (_shouldRun && _clientId == clientId) return;
            StopLocked(clearFirst: false);
            _clientId = clientId;
            _shouldRun = true;
            _cts = new CancellationTokenSource();
            _loop = Task.Run(() => RunAsync(_cts.Token));
        }
    }

    /// <summary>Stop reconnecting, clear the presence, and close the pipe.</summary>
    public void Disconnect()
    {
        lock (_lock)
        {
            _currentActivity = null;
            StopLocked(clearFirst: true);
            SetState(ConnState.Disconnected);
        }
    }

    /// <summary>Set (or with null, clear) the presence. Deduplicates: an
    /// unchanged payload is not re-sent.</summary>
    public void SetActivity(RichPresence? activity)
    {
        JsonNode? node = activity is null ? null : JsonSerializer.SerializeToNode(activity, WireOptions);
        var signature = node?.ToJsonString() ?? "CLEARED";

        NamedPipeClientStream? pipe;
        lock (_lock)
        {
            _currentActivity = node;
            if (!_ready || signature == _sentSignature) return;
            _sentSignature = signature;
            pipe = _pipe;
        }
        if (pipe is null) return;
        _ = Task.Run(async () =>
        {
            try { await SendActivityAsync(pipe, node, CancellationToken.None); }
            catch { /* the read loop notices the broken pipe and reconnects */ }
        });
    }

    /// <summary>Best-effort synchronous clear, used during app termination so
    /// the presence does not get stuck. Returns quickly either way.</summary>
    public void ClearActivitySync()
    {
        NamedPipeClientStream? pipe;
        lock (_lock)
        {
            if (!_ready) return;
            pipe = _pipe;
        }
        if (pipe is null) return;
        try { SendActivityAsync(pipe, null, CancellationToken.None).Wait(500); }
        catch { }
    }

    public void Dispose()
    {
        lock (_lock) StopLocked(clearFirst: true);
    }

    // --- Connection lifecycle

    /// <summary>Must hold _lock. Cancels the loop and closes the pipe.</summary>
    private void StopLocked(bool clearFirst)
    {
        _shouldRun = false;
        if (clearFirst && _ready && _pipe is { } p)
        {
            try { SendActivityAsync(p, null, CancellationToken.None).Wait(500); }
            catch { }
        }
        _cts?.Cancel();
        _cts = null;
        _loop = null;
        ClosePipeLocked();
    }

    private void ClosePipeLocked()
    {
        _ready = false;
        _sentSignature = null;
        try { _pipe?.Dispose(); } catch { }
        _pipe = null;
    }

    private async Task RunAsync(CancellationToken ct)
    {
        var attempt = 0;
        while (!ct.IsCancellationRequested)
        {
            SetState(ConnState.Connecting);
            var pipe = OpenFirstAvailablePipe();
            if (pipe is null)
            {
                SetState(ConnState.Disconnected);
                if (!await Backoff(++attempt, ct)) return;
                continue;
            }

            lock (_lock)
            {
                _pipe = pipe;
                _ready = false;
                _sentSignature = null;
            }

            try
            {
                var handshake = JsonSerializer.SerializeToUtf8Bytes(
                    new { v = 1, client_id = _clientId });
                await WriteFrameAsync(pipe, Opcode.Handshake, handshake, ct);
                SetState(ConnState.Connected);
                await ReadLoopAsync(pipe, ct);
            }
            catch (OperationCanceledException)
            {
                return;
            }
            catch
            {
                // Fall through to reconnect.
            }

            lock (_lock) ClosePipeLocked();
            if (ct.IsCancellationRequested) return;
            SetState(ConnState.Disconnected);
            if (!await Backoff(++attempt, ct)) return;
        }
    }

    /// <summary>Reads frames until the pipe drops. Handles READY (mark ready,
    /// re-send the current activity), PING (reply PONG), ERROR, and CLOSE.</summary>
    private async Task ReadLoopAsync(NamedPipeClientStream pipe, CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            var frame = await ReadFrameAsync(pipe, ct);
            if (frame is null) return; // peer closed
            var (opcode, payload) = frame.Value;

            switch ((Opcode)opcode)
            {
                case Opcode.Frame:
                    JsonDocument doc;
                    try { doc = JsonDocument.Parse(payload); }
                    catch { continue; }
                    using (doc)
                    {
                        var evt = doc.RootElement.TryGetProperty("evt", out var e) && e.ValueKind == JsonValueKind.String
                            ? e.GetString() : null;
                        if (evt == "READY")
                        {
                            JsonNode? resend;
                            lock (_lock)
                            {
                                _ready = true;
                                resend = _currentActivity;
                                _sentSignature = resend?.ToJsonString() ?? "CLEARED";
                            }
                            Ready?.Invoke();
                            try { await SendActivityAsync(pipe, resend, ct); } catch { return; }
                        }
                        else if (evt == "ERROR")
                        {
                            var message = doc.RootElement.TryGetProperty("data", out var d)
                                && d.ValueKind == JsonValueKind.Object
                                && d.TryGetProperty("message", out var m)
                                && m.ValueKind == JsonValueKind.String
                                ? m.GetString() : null;
                            Error?.Invoke(message ?? "Discord reported an error");
                        }
                    }
                    break;

                case Opcode.Ping:
                    await WriteFrameAsync(pipe, Opcode.Pong, payload, ct);
                    break;

                case Opcode.Close:
                    return;
            }
        }
    }

    private static async Task<bool> Backoff(int attempt, CancellationToken ct)
    {
        // Exponential backoff capped at 30s, matching the macOS client.
        var delay = Math.Min(Math.Pow(2, Math.Min(attempt, 5)), 30.0);
        try { await Task.Delay(TimeSpan.FromSeconds(delay), ct); }
        catch (OperationCanceledException) { return false; }
        return true;
    }

    // --- Framing

    private async Task SendActivityAsync(NamedPipeClientStream pipe, JsonNode? activity, CancellationToken ct)
    {
        // The args object must carry an explicit "activity": null to clear the
        // presence, so it is built as a JsonObject rather than serialized from a
        // model with null-omitting options.
        var command = new JsonObject
        {
            ["cmd"] = "SET_ACTIVITY",
            ["nonce"] = Guid.NewGuid().ToString(),
            ["args"] = new JsonObject
            {
                ["pid"] = _pid,
                ["activity"] = activity?.DeepClone(),
            },
        };
        await WriteFrameAsync(pipe, Opcode.Frame, Encoding.UTF8.GetBytes(command.ToJsonString()), ct);
    }

    private async Task WriteFrameAsync(NamedPipeClientStream pipe, Opcode opcode, byte[] payload, CancellationToken ct)
    {
        var frame = new byte[8 + payload.Length];
        BitConverter.TryWriteBytes(frame.AsSpan(0, 4), (uint)opcode);
        BitConverter.TryWriteBytes(frame.AsSpan(4, 4), (uint)payload.Length);
        payload.CopyTo(frame, 8);

        // Serialize writes: the read loop's PONGs and SetActivity calls would
        // otherwise interleave frames.
        await _writeLock.WaitAsync(ct);
        try
        {
            await pipe.WriteAsync(frame, ct);
            await pipe.FlushAsync(ct);
        }
        finally
        {
            _writeLock.Release();
        }
    }

    /// <summary>Read one frame; null when the peer closed the pipe.</summary>
    private static async Task<(uint Opcode, byte[] Payload)?> ReadFrameAsync(NamedPipeClientStream pipe, CancellationToken ct)
    {
        var header = await ReadExactAsync(pipe, 8, ct);
        if (header is null) return null;
        var opcode = BitConverter.ToUInt32(header, 0);
        var length = BitConverter.ToUInt32(header, 4);
        if (length == 0) return (opcode, []);
        var payload = await ReadExactAsync(pipe, (int)length, ct);
        return payload is null ? null : (opcode, payload);
    }

    private static async Task<byte[]?> ReadExactAsync(NamedPipeClientStream pipe, int count, CancellationToken ct)
    {
        var buffer = new byte[count];
        var total = 0;
        while (total < count)
        {
            int n;
            try { n = await pipe.ReadAsync(buffer.AsMemory(total, count - total), ct); }
            catch (IOException) { return null; }
            if (n <= 0) return null;
            total += n;
        }
        return buffer;
    }

    // --- Pipe discovery

    /// <summary>Open the first available \\.\pipe\discord-ipc-{0..9}.</summary>
    private static NamedPipeClientStream? OpenFirstAvailablePipe()
    {
        for (var i = 0; i <= 9; i++)
        {
            var pipe = new NamedPipeClientStream(
                ".", $"discord-ipc-{i}", PipeDirection.InOut, PipeOptions.Asynchronous);
            try
            {
                pipe.Connect(200);
                return pipe;
            }
            catch
            {
                pipe.Dispose();
            }
        }
        return null;
    }

    // --- Helpers

    private void SetState(ConnState newState)
    {
        if (State == newState) return;
        State = newState;
        StateChanged?.Invoke(newState);
    }
}
